### PR TITLE
Refactored matrices from lanczos and jacobi into matrix.c

### PR DESCRIPTION
**Pull request by felix642**
Extracted methods related to matrices and vectors from lanczos.c and jacobi.c and combined them in the file matrix.c

### CODE REVIEW

File: src/matrix.c [hunk 1]
```diff
@@ -0,0 +1,72 @@
+#include "matrix.h"
+
+#include <math.h>
+
```
Comment (GM07) [inline src/matrix.c:4]: what happened to the documentation of those functions ?

File: src/matrix.c [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    for (matrix_size i = 0; i < sizeSquared; ++i) {
+        if (isDiagonal(i, size)) {
+            output[i] = 1;
+        } else {
+            output[i] = 0;
+        }
+    }
+}
+
+void transpose(matrix_real_number* input, matrix_real_number* output, matrix_size* dims) {
```
Comment (GM07) [inline src/matrix.c:39]: I would name that function `matrixTranspose` to respect the convention defined by the other functions or I would rename other functions to respect that convention (for example simply put `multiply` for the multiplication function

File: src/matrix.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Multiplies 2 matrices together using this equation : output = firstMatrix x secondMatrix
+ * @param firstMatrix first matrix to multiply of size m * n
+ * @param secondMatrix second matrix to multiply of size n * p
+ * @param size array of size 3 containing m, n and p
+ * @param output matrix resulting of the matrix multiplication of the given matrices
+ * @param transposeFirstMatrix whether to transpose the first matrix when multiplying the matrices
+ */
+void matrixMultiply(const matrix_real_number* firstMatrix, const matrix_real_number* secondMatrix, const matrix_size size[3], matrix_real_number* output, const matrix_size transposeFirstMatrix);
```
Comment (GM07) [inline src/matrix.h:25]: I just saw that the documentation of the functions are in the .h file. We always put them in the .c file so I would move that to the .c file. Also, there are certain functions without a definition.

File: src/matrix.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+typedef unsigned int matrix_size; 
+typedef double matrix_real_number;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Converts a 2d coordinate in a matrix to an index in a 2d array
+*/
```
Comment (nicholaslegrand102) [inline src/matrix.h:13]: Couldn't this be a // comment?

File: src/matrix.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * @param firstMatrix first matrix to multiply of size m * n
+ * @param secondMatrix second matrix to multiply of size n * p
+ * @param size array of size 3 containing m, n and p
+ * @param output matrix resulting of the matrix multiplication of the given matrices
+ * @param transposeFirstMatrix whether to transpose the first matrix when multiplying the matrices
+ */
+void matrixMultiply(const matrix_real_number* firstMatrix, const matrix_real_number* secondMatrix, const matrix_size size[3], matrix_real_number* output, const matrix_size transposeFirstMatrix);
+
+/**
+ * Creates the identity matrix (1 on the diagonal and 0 on other coordinates)
```
Comment (nicholaslegrand102) [inline src/matrix.h:28]: This should be a @brief

**Review by nicholaslegrand102 — APPROVED** [review]
Looks good to me

File: src/matrix.c [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+ * @param output output matrix
+ */
+void createIdentityMatrix(const matrix_size size, matrix_real_number* output) {
+    const matrix_size sizeSquared = size * size;
+    for (matrix_size i = 0; i < sizeSquared; ++i) {
+        if (isDiagonal(i, size)) {
+            output[i] = 1;
+        } else {
+            output[i] = 0;
+        }
```
Comment (zedfly) [inline src/matrix.c:63]: Why no use the tertiary operator here it seems like it would make the code more readable

File: src/matrix.c [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+ * @param input the matrix to transpose
+ * @param output an array to store the transposed matrix
+ * @param dims an array of size 3 with the number of rows of the first matrix,
+ * the number of rows of the second matrix and the number of columns of the
+ * second matrix
+ */
+void matrixTranspose(matrix_real_number* input, matrix_real_number* output, matrix_size* dims) {
+  for (matrix_size i = 0; i < dims[0]; ++i) {
+    for (matrix_size j = 0; j < dims[1]; ++j) {
+      output[j * dims[0] + i] = input[i * dims[1] + j];
```
Comment (zedfly) [inline src/matrix.c:78]: We could use coordsToIndex here

File: src/matrix.c [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+ * firstMatrix x secondMatrix
+ * @param firstMatrix first matrix to multiply of size m * n
+ * @param secondMatrix second matrix to multiply of size n * p
+ * @param size array of size 3 containing m, n and p
+ * @param output matrix resulting of the matrix multiplication of the given
+ * matrices
+ * @param transposeFirstMatrix whether to transpose the first matrix when
+ * multiplying the matrices
+ */
+void matrixMultiply(const matrix_real_number* firstMatrix,
```
Comment (bisson2000) [inline src/matrix.c:35]: This looks like a cross product. Is it one? If so, we could define this as one. Having it defined as multiply could make others confused as whether this is a dot product or a cross product.

**Review by bisson2000 — COMMENTED** [review]
I left a comment, you should check it out

File: src/matrix.c (hunk 8 above)
Comment (felix642) [reply src/matrix.c:35]: Cross product and dot product are only applicable to vectors. In this case we are working with matrices and those terms do not apply. I do not think such change is desirable.

**Review by bisson2000 — APPROVED** [review]
gfm

**Review by zedfly — COMMENTED** [review]
LGTM

**Review by zedfly — APPROVED** [review]

**Pull request merged**